src/librustc_traits/chalk/lowering.rs

```diff
@@ -884,3 +993,95 @@ impl<'tcx> TypeFolder<'tcx> for ParamsSubstitutor<'tcx> {
         }
     }
 }
+
+/// Used to collect `Placeholder`s.
+crate struct PlaceholdersCollector {
+    universe_index: ty::UniverseIndex,
+    crate next_ty_placehoder: usize,
+    crate next_anon_region_placeholder: u32,
+}
+
+impl PlaceholdersCollector {
+    crate fn new() -> Self {
+        PlaceholdersCollector {
+            universe_index: ty::UniverseIndex::ROOT,
+            next_ty_placehoder: 0,
+            next_anon_region_placeholder: 0,
+        }
+    }
+}
+
+impl<'tcx> TypeVisitor<'tcx> for PlaceholdersCollector {
+    fn visit_binder<T: TypeFoldable<'tcx>>(&mut self, t: &Binder<T>) -> bool {
+        t.super_visit_with(self)
+    }
+
+    fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+        match t.kind {
+            ty::Placeholder(p) if p.universe == self.universe_index => {
+                self.next_ty_placehoder = self.next_ty_placehoder.max(p.name.as_usize() + 1);
+            }
+
+            _ => (),
+        };
+
+        t.super_visit_with(self)
```

Comment:
Can be moved to the wildcard match, I think? Same with `visit_region`

Reply:
Mmm no, because this returns `bool`